packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol

```diff
@@ -0,0 +1,234 @@
+                revert(0x1C, 0x04)
+            }
+        }
+        _;
+    }
+
+    /// @notice Retrieves the sender of the current cross domain message. If not entered, reverts.
+    /// @return _sender Address of the sender of the current cross domain message.
+    function crossDomainMessageSender() external view notEntered returns (address _sender) {
+        assembly {
+            _sender := tload(CROSS_DOMAIN_MESSAGE_SENDER_SLOT)
+        }
+    }
+
+    /// @notice Retrieves the source of the current cross domain message. If not entered, reverts.
+    /// @return _source Chain ID of the source of the current cross domain message.
+    function crossDomainMessageSource() external view notEntered returns (uint256 _source) {
+        assembly {
+            _source := tload(CROSS_DOMAIN_MESSAGE_SOURCE_SLOT)
+        }
+    }
+
+    /// @notice Sends a message to some target address on a destination chain. Note that if the call always reverts,
+    ///         then the message will be unrelayable and any ETH sent will be permanently locked. The same will occur
+    ///         if the target on the other chain is considered unsafe (see the _isUnsafeTarget() function).
+    /// @param _destination Chain ID of the destination chain.
+    /// @param _target      Target contract or wallet address.
+    /// @param _message     Message payload to call target with.
+    function sendMessage(uint256 _destination, address _target, bytes calldata _message) external payable {
+        if (_destination == block.chainid) revert MessageDestinationSameChain();
+        if (_target == Predeploys.CROSS_L2_INBOX) revert MessageTargetCrossL2Inbox();
+        if (_target == Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER) revert MessageTargetL2ToL2CrossDomainMessenger();
+
+        bytes memory data = abi.encodeCall(
+            L2ToL2CrossDomainMessenger.relayMessage,
+            (_destination, block.chainid, messageNonce(), msg.sender, _target, _message)
+        );
+        emit SentMessage(data);
+        msgNonce++;
+    }
+
+    /// @notice Relays a message that was sent by the other CrossDomainMessenger contract. Can only be executed via
+    ///         cross-chain call from the other messenger OR if the message was already received once and is currently
+    ///         being replayed.
+    /// @param _destination Chain ID of the destination chain.
+    /// @param _source      Chain ID of the source chain.
+    /// @param _nonce       Nonce of the message being relayed.
+    /// @param _sender      Address of the user who sent the message.
+    /// @param _target      Address that the message is targeted at.
+    /// @param _message     Message payload to call target with.
+    function relayMessage(
+        uint256 _destination,
+        uint256 _source,
+        uint256 _nonce,
+        address _sender,
+        address _target,
+        bytes memory _message
+    )
+        external
+        payable
+        nonReentrant
```

Comment:
We want to define our own reentrancy guard, in pseudocode

```solidity
modifier nonReentrant {
  if eq(tload(slot), 1) {
    // mstore, revert
  }
  tstore(slot, 1)
  _;
  tstore(slot, 0)
}

```

Reply:
could we use `ENTERED_SLOT` for slot?

Reply:
If you use `ENTERED_SLOT` then it works great